### PR TITLE
Add MDS050 proper-names rule for enforcing proper name casing

### PR DESCRIPTION
**Pull request by jeduden**
## Summary

Implements MDS050, a new linting rule that enforces consistent casing for configured proper names (e.g., `JavaScript`, `GitHub`, `mdsmith`) in Markdown documents.

## Key Changes

- **New rule implementation** (`internal/rules/propernames/rule.go`):
  - Scans Markdown prose for case-insensitive matches of configured proper names
  - Enforces whole-word matching with left-boundary detection (word characters: `a-z`, `A-Z`, `0-9`, `_`)
  - Implements `Check()` to detect wrong-cased occurrences and emit diagnostics
  - Implements `Fix()` to replace wrong-cased matches with canonical spellings
  - Supports optional checking inside code spans/blocks and HTML via `check-code` and `check-html` settings
  - Implements `rule.Configurable` with three settings:
    - `names`: list of proper names (appends across config layers)
    - `check-code`: bool to check code blocks (replaces)
    - `check-html`: bool to check HTML blocks (replaces)
  - Disabled by default (`EnabledByDefault()` returns `false`)

- **Comprehensive test suite** (`internal/rules/propernames/rule_test.go`):
  - Tests for correct/wrong casing detection
  - Tests for word boundary edge cases (`JavaScripts`, `GitHubber`)
  - Tests for different Markdown contexts (headings, links, code spans)
  - Tests for configuration and settings merging
  - Tests for fix functionality

- **Documentation and fixtures** (`internal/rules/MDS050-proper-names/`):
  - README with rule description, settings, detection logic, and examples
  - Fixture tests covering bad cases (wrong casing in prose, headings, links, code blocks)
  - Fixture tests covering good cases (correct casing, word boundaries, skipped code)
  - Fixed output examples

- **Integration**:
  - Registered rule in `cmd/mdsmith/main.go` and `internal/integration/rules_test.go`
  - Updated rule index and plan tracking

## Notable Implementation Details

- **Word boundary matching**: Uses a simple ASCII word character check (`isWordChar`) to ensure matches only occur at word boundaries, preventing false positives in compound words
- **Case-insensitive scanning**: Converts both the text and configured names to lowercase for comparison, then validates actual casing against the canonical form
- **AST traversal**: Walks the Markdown AST, skipping code/HTML nodes by default but allowing them to be checked via settings
- **Safe rewriting**: The `Fix()` method sorts matches by offset and applies replacements sequentially to avoid overlapping rewrites
- **Append merge mode**: The `names` setting uses `MergeAppend` to allow kind-layer configurations to extend the inherited vocabulary rather than replace it

https://claude.ai/code/session_01K9MD5XeGkXccnX6EbK4L71

### CODE REVIEW

File: internal/rules/propernames/rule.go [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+		for i := 0; i <= len(lowerText)-n; i++ {
+			// Left boundary: the byte before the match (in source) must not
+			// be a word character, or the match is at the start of the source.
+			absOffset := baseOffset + i
+			if absOffset > 0 && isWordChar(source[absOffset-1]) {
+				continue
+			}
+			// Case-insensitive prefix match.
+			if !bytes.Equal(lowerText[i:i+n], []byte(lowerName)) {
+				continue
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in ba3b829. `lowerText` is now computed once before the names loop, and `lowerName` is a `[]byte` allocated once per name rather than once per position.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+	var out bytes.Buffer
+	prev := 0
+	for _, m := range matches {
+		if m.start < prev {
+			continue // overlapping match (shouldn't happen but guard anyway)
+		}
+		out.Write(f.Source[prev:m.start])
+		out.WriteString(m.name)
+		prev = m.start + m.length
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in ba3b829. The sort comparator now breaks ties by `matches[i].length > matches[j].length`, so at the same start offset the longest match wins and shorter overlapping entries are subsequently skipped. A test (`TestFix_OverlappingMatches_LongestMatchWins`) verifies this.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+			}
+			return ast.WalkSkipChildren, nil
+		case *ast.FencedCodeBlock, *ast.CodeBlock:
+			if r.CheckCode {
+				all = append(all, r.scanLines(n, f)...)
+			}
+			return ast.WalkSkipChildren, nil
+		case *ast.HTMLBlock:
+			if r.CheckHTML {
+				all = append(all, r.scanLines(n, f)...)
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: This compiles correctly. In goldmark, `ast.Node` is a Go interface that includes `Lines() *text.Segments` (see goldmark's `ast/node.go`). Our `lineNode` interface is a structural subset of `ast.Node`, so the multi-type switch case passes `n` (type `ast.Node`) to `scanLines` without any type assertion needed. `go build ./...` succeeds and all tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Same as the FencedCodeBlock/CodeBlock case: goldmark's `ast.Node` interface includes `Lines() *text.Segments`, so `n` (type `ast.Node`) satisfies `lineNode` and can be passed to `scanLines` directly. The code compiles and tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: This compiles. goldmark's `ast.Node` interface (defined in `goldmark/ast/node.go`) includes `Lines() *text.Segments`, so any value of type `ast.Node` already satisfies our `lineNode` interface. No type assertion is needed. `go build ./...` and all tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Same as above — goldmark's `ast.Node` interface includes `Lines() *text.Segments`, so `n` (type `ast.Node`) satisfies `lineNode` directly. The code compiles and tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+			Line:     f.LineOfOffset(m.start),
+			Column:   f.ColumnOfOffset(m.start),
+			RuleID:   r.ID(),
+			RuleName: r.Name(),
+			Severity: lint.Warning,
+			Message:  fmt.Sprintf("proper name %q should be %q", m.actual, m.name),
+		})
+	}
+	return diags
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go:273]: Fixed in 227fae2. Extracted a `normalizeMatches` helper that sorts by offset (longest-match-wins tiebreak) and removes overlapping entries; both `Check` and `Fix` now call it. A new test (`TestCheck_OverlappingMatches_OneDiagnostic`) verifies `Check` emits exactly one diagnostic for an overlapping span.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+// Fix implements rule.FixableRule. It replaces each wrong-cased match with
+// the canonical spelling in place. Whole-word left-boundary matching makes
+// this a safe rewrite.
+func (r *Rule) Fix(f *lint.File) []byte {
+	matches := r.collectMatches(f)
+	if len(matches) == 0 {
+		out := make([]byte, len(f.Source))
+		copy(out, f.Source)
+		return out
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go:285]: No change needed here. The engine's `check.go` already calls `filterGeneratedDiags(diags, f.GeneratedRanges)` after every rule's `Check`, so generated-section content is excluded from diagnostics. For `Fix`, the same pattern applies to all fixable text-style rules in this codebase (e.g. `notrailingspaces`, `linelength`) — none filter generated ranges at the rule level; that responsibility sits with the engine. Adding generated-range filtering to this rule alone would be inconsistent.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+			Severity: lint.Warning,
+			Message:  fmt.Sprintf("proper name %q should be %q", m.actual, m.name),
+		})
+	}
+	return diags
+}
+
+// Fix implements rule.FixableRule. It replaces each wrong-cased match with
+// the canonical spelling in place. Whole-word left-boundary matching makes
+// this a safe rewrite.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in db46fa8. The comment now reads: "Only the left boundary is enforced — the byte before the match must be a non-word character (or start of file), but no right-boundary check is applied. The matched prefix is replaced and any trailing word characters (e.g. the 's' in 'JavaScripts') are left as-is."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+// inherited vocabulary rather than replace it.
+func (r *Rule) ApplySettings(s map[string]any) error {
+	for k, v := range s {
+		switch k {
+		case "names":
+			names, ok := settings.ToStringSlice(v)
+			if !ok {
+				return fmt.Errorf("proper-names: names must be a list of strings, got %T", v)
+			}
+			r.Names = names
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go:310]: Not adding validation here. Configuring two names that are identical under case-folding (e.g. `["Go", "go"]`) is a contradictory user config — any occurrence of "go" already has the correct casing for one of the two names, so in practice only the other produces a diagnostic. For an all-caps occurrence like "GO", both match at equal length and `normalizeMatches` keeps one non-deterministically; the result is still a valid canonical replacement from the configured vocabulary. Adding ApplySettings validation would introduce complexity (and a new error path to test) for an edge case that only arises from a configuration error. No other rule validates semantic consistency of its settings list.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/README.md [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+```
+
+## Detection
+
+For each configured name, the rule scans prose text for
+case-insensitive matches that start at a word boundary. The byte
+immediately before the match in the full source must be a
+non-alphanumeric, non-underscore character, or the match must
+begin at the start of the file. When the matched bytes differ from
+the canonical spelling, a diagnostic is emitted.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/README.md]: Fixed in db46fa8. The Detection section now reads: "Only the **left** boundary is enforced — there is no right-boundary check. A name like `JavaScript` therefore matches the prefix of `JavaScripts` and reports (and fixes) that prefix, leaving the trailing `s` unchanged."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: plan/126_proper-names.md [hunk 10]
```diff
@@ -140,32 +140,32 @@ proper name "{actual}" should be "{configured}"
...
+- [x] `JavaScript is fun` with `names: [JavaScript]`
       emits no diagnostic.
-- [ ] `Javascript is fun` emits one diagnostic and
+- [x] `Javascript is fun` emits one diagnostic and
       fixes to `JavaScript is fun`.
-- [ ] `JAVASCRIPT` emits one diagnostic.
-- [ ] `JavaScripts` (trailing letter) emits one
+- [x] `JAVASCRIPT` emits one diagnostic.
+- [x] `JavaScripts` (trailing letter) emits one
       diagnostic on the `JavaScript` portion.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply plan/126_proper-names.md]: Fixed in ce20bb0. The criterion now reads: "`Javascripts` (trailing letter, wrong case) emits one diagnostic on the `Javascript` portion."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/README.md [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+## Detection
+
+For each configured name, the rule scans prose text for
+case-insensitive matches that start at a word boundary. The byte
+immediately before the match in the full source must be a
+non-alphanumeric, non-underscore character, or the match must
+begin at the start of the file. Only the **left** boundary is
+enforced — there is no right-boundary check. A name like
+`JavaScript` therefore matches the prefix of `JavaScripts` and
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/README.md]: Fixed in ce20bb0. The Detection section now uses `Javascripts` (wrong case) as the example and adds: "Correctly-cased prefixes in longer words (e.g. `JavaScripts`) are not flagged."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+	lowerText := bytes.ToLower(text)
+	var results []wrongMatch
+	for _, name := range r.Names {
+		n := len(name)
+		if n == 0 || n > len(text) {
+			continue
+		}
+		// Lowercase the name once per name, not once per position.
+		lowerName := []byte(strings.ToLower(name))
+		for i := 0; i <= len(lowerText)-n; i++ {
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go:116]: Go's `bytes.ToLower` is UTF-8-aware (it uses `unicode.ToLower` on each decoded rune), so the "per-byte corruption" concern doesn't apply. The length-mismatch concern is real in theory: if `strings.ToLower(name)` produces a byte string of different length than `len(name)` (e.g. for Turkish İ or German ß), the `bytes.Equal(lowerText[i:i+n], lowerName)` comparison would always fail because the slice lengths differ. In practice this rule is designed for ASCII product names (JavaScript, GitHub, TypeScript), where `strings.ToLower` never changes byte length. No change to the implementation; the README now explicitly states that boundary detection uses ASCII `[A-Za-z0-9_]`, which sets clear expectations about the rule's scope.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/README.md [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+immediately before the match in the full source must be a
+non-alphanumeric, non-underscore character, or the match must
+begin at the start of the file. Only the **left** boundary is
+enforced — there is no right-boundary check. For example,
+`Javascripts` (wrong case) matches `JavaScript` at the left
+boundary; the rule reports and fixes the `Javascripts` prefix to
+`JavaScript`, leaving the trailing `s` unchanged. Correctly-cased
+prefixes in longer words (e.g. `JavaScripts`) are not flagged.
+When the matched bytes differ from the canonical spelling, a
+diagnostic is emitted.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/README.md]: Fixed in 14bff93. The Detection section now reads: "must be a non-ASCII-letter, non-ASCII-digit, non-underscore character (i.e. not in `[A-Za-z0-9_]`)".

---
_Generated by [Claude Code](https://claude.ai/code)_

File: cmd/mdsmith/main.go [hunk 14]
```diff
@@ -61,6 +61,7 @@ import (
 	_ "github.com/jeduden/mdsmith/internal/rules/orderedlistnumbering"
 	_ "github.com/jeduden/mdsmith/internal/rules/paragraphreadability"
 	_ "github.com/jeduden/mdsmith/internal/rules/paragraphstructure"
+	_ "github.com/jeduden/mdsmith/internal/rules/propernames"
 	_ "github.com/jeduden/mdsmith/internal/rules/recipesafety"
 	_ "github.com/jeduden/mdsmith/internal/rules/requiredstructure"
```
[the remark this answers is not in the document]
Comment (jeduden) [reply cmd/mdsmith/main.go:66]: Fixed in 4697822. Added `_ "github.com/jeduden/mdsmith/internal/rules/propernames"` to both `internal/config/config_test.go` and `internal/engine/categories_test.go` in alphabetical order. Also added `"prose"` to `ValidCategories` in `internal/config/config.go` so the `TestAllRuleCategoriesValid` check passes.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 15]
```diff
@@ -0,0 +1,317 @@
...
+			if absOffset > 0 && isWordChar(source[absOffset-1]) {
+				continue
+			}
+			// Case-insensitive prefix match.
+			if !bytes.Equal(lowerText[i:i+n], lowerName) {
+				continue
+			}
+			// Compare actual casing to the canonical name.
+			actual := string(text[i : i+n])
+			if actual == name {
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in 4697822. Replaced `bytes.ToLower` and `strings.ToLower` with a new `asciiToLower` helper that only folds `A-Z → a-z` and never changes byte length, eliminating the offset/panic concern entirely. The loop bound and `text[i:i+n]` slice now operate on identical-length arrays.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/README.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+non-ASCII-letter, non-ASCII-digit, non-underscore character (i.e.
+not in `[A-Za-z0-9_]`), or the match must begin at the start of
+the file. Only the **left** boundary is enforced — there is no
+right-boundary check. For example, `Javascripts` (wrong case)
+matches `JavaScript` at the left boundary; the rule reports and
+fixes the `Javascripts` prefix to `JavaScript`, leaving the
+trailing `s` unchanged. Correctly-cased prefixes in longer words
+(e.g. `JavaScripts`) are not flagged.
+When the matched bytes differ from the canonical spelling, a
+diagnostic is emitted.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/README.md]: Fixed in 0e78c9e. The Detection section now reads "ASCII-case-insensitive matches" and adds: "Only ASCII letters (`A`–`Z`) are folded; non-ASCII characters in a configured name are matched byte-for-byte."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/README.md [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/README.md]: Fixed in a98603c. The sentence now reads: "A match must be preceded by a non-word character (not in `[A-Za-z0-9_]`) or occur at the start of the file."

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule_test.go [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+	require.Len(t, diags, 1)
+	assert.Equal(t, `proper name "JAVASCRIPT" should be "JavaScript"`, diags[0].Message)
+}
+
+func TestCheck_Fix_ReplacesWrongCasing(t *testing.T) {
+	f := newFile(t, "Use javascript and github.\n")
+	r := &Rule{Names: []string{"JavaScript", "GitHub"}}
+	fixed := r.Fix(f)
+	assert.Equal(t, "Use JavaScript and GitHub.\n", string(fixed))
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule_test.go]: Fixed in a98603c. Renamed to `TestFix_ReplacesWrongCasing`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 19]
```diff
@@ -0,0 +1,332 @@
...
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in f3565c7. Replaced `string(text[i:i+n])` with `bytes.Equal(text[i:i+n], nameBytes)`, so correct-casing matches (the common case) never allocate a string. The `actual` string is now computed lazily in `Check` only when a diagnostic is actually emitted.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 20]
```diff
@@ -0,0 +1,332 @@
...
+}
+
+// wrongMatch holds one wrong-cased occurrence.
+type wrongMatch struct {
+	start  int // byte offset in f.Source
+	length int
+	actual string
+	name   string
+}
+
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go]: Fixed in f3565c7. Removed `actual string` from `wrongMatch`. `Fix` never used it, and `Check` now computes the actual string lazily from `f.Source[m.start:m.start+m.length]` only when building the diagnostic message.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/propernames/rule.go [hunk 21]
```diff
@@ -0,0 +1,330 @@
...
+	var results []wrongMatch
+	for _, name := range r.Names {
+		n := len(name)
+		if n == 0 || n > len(text) {
+			continue
+		}
+		// Lowercase the name once per name, not once per position.
+		nameBytes := []byte(name)
+		lowerName := asciiToLower(nameBytes)
+		for i := 0; i <= len(lowerText)-n; i++ {
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/propernames/rule.go:116]: Fixed in 73fb036. Introduced a `nameEntry` type holding the precomputed `canonical []byte` and `lower []byte` for each name. `buildNameEntries()` runs once in `collectMatches` and the resulting slice is passed to `scanBytes`/`scanLines`/`scanCodeSpanChildren`/`scanRawHTMLSegments`, so no per-segment allocations occur for these values.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS050-proper-names/bad/check-code-enabled.md [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+diagnostics:
+  - line: 4
+    column: 1
+    message: 'proper name "javascript" should be "JavaScript"'
+---
+# Code Block Checked
+
+```
+javascript --version
+```
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS050-proper-names/bad/check-code-enabled.md:15]: Fixed in 73fb036. Added `fixed/check-code-enabled.md` with `JavaScript --version` as the expected fix output. Also added a `sh` language tag to the `bad/check-code-enabled.md` fenced block so the all-rules fixed-output check (MDS011) passes.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by jeduden** [comment]
🟢 **Merge Queue** — picked up

This PR is in the queue and will be batched with other `queue`-labelled PRs.

**Next:** No action needed — you'll get another comment when CI starts on the batch. [View merge queue run](https://github.com/jeduden/mdsmith/actions/runs/25293791142).

**Comment by jeduden** [comment]
🔵 **Merge Queue** — CI running

Merged into batch branch [`merge-queue/batch-227-1777850546`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-227-1777850546). [View CI run](https://github.com/jeduden/mdsmith/actions/runs/25293796760).

**Next:** No action needed — you'll be notified when CI completes.

**Comment by jeduden** [comment]
✅ **Merge Queue** — merged

This PR landed on `main` via commit [`c6cbfd9`](https://github.com/jeduden/mdsmith/commit/c6cbfd9a1d884df15c9a940dc1726681127f5974). [CI run that validated the merge](https://github.com/jeduden/mdsmith/actions/runs/25293796760).

**Next:** Done — nothing more to do here.
